File: array/33.py
from typing import List

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        if len(nums) == 0: return -1
        low = 0
        high = len(nums) - 1
        while low <= high:
            mid = int((low + high) / 2)
            if nums[mid] == target: return mid
            if nums[low] == target: return low
            if nums[high] == target: return high

            if nums[low] < nums[mid]:
                # low 到 mid 有序
                if nums[low] > target or nums[mid] < target:
                    low = mid + 1
                    continue
                else:
                    return self.binarySearch(nums, low, mid, target)
            else:
                #low 到 mid 无序， mid 到 high必定有序
                if nums[mid] > target or nums[high] < target:
                    high = mid - 1
                else:
                    return self.binarySearch(nums, mid, high, target)
        return -1



    def binarySearch(self, nums, low, high, target):
        while low <= high:
            mid = int((low + high) / 2)
            if nums[mid] < target:
                low = mid + 1
            elif nums[mid] > target:
                high = mid - 1
            else:
                return mid
        return -1
        



def main():
    a = [4,5,6,7,0,1,2]
    val = 3
    print(Solution().search(a, val))

if __name__ == "__main__":
    main()

File: leetcode-cpp/33.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int search(vector<int>& nums, int target) {
        int l = 0, r = nums.size() - 1;
        int m = l + (r - l) / 2;
        while(l <= r) {
            if (nums[m] == target){
                return m;
            }
            if (nums[l] <= nums[m]) {
                // 左侧有序
                if (nums[l] <= target && nums[m] >= target)
                    r = m - 1;
                else
                    l = m + 1;
            } else {
                // 右侧有序
                if (nums[m] <= target && nums[r] >= target)
                    l = m + 1;
                else
                    r = m - 1;
            }
            m = l + (r - l) / 2;
        }
        return -1;
    }
};

int main(){
    Solution s;
    vector<int> nums({4,5,6,7,0,1,2});
    int target = 3;
    int ret = s.search(nums, target);
    cout << ret << endl;
}
